File: ast/expression/identifier.go
package expression

import (
	"fmt"

	"github.com/panda-io/micro-panda/ast/ast"
)

type Identifier struct {
	ExpressionBase
	Name string

	Qualified   string
	IsNamespace bool
}

func (i *Identifier) Validate(c ast.Context, expected ast.Type) {
	t := c.FindObject(i.Name)
	if t == nil {
		d := c.FindLocalDeclaration(i.Name)
		if d == nil {
			i.IsNamespace = c.IsNamespace(i.Name)
		} else if d.Kind() != ast.DeclarationStruct {
			i.Const = d.IsConstant()
			i.Typ = d.Type()
			i.Qualified = d.QualifiedName()
		}
	} else {
		i.Const = false
		i.Typ = t
	}
	// * type would be nil for enum (its member has type u8)
	// * type is nil when identifier is namespacee
	if i.Typ == nil && i.Qualified == "" && !i.IsNamespace {
		c.Error(i.GetPosition(), fmt.Sprintf("undefined %s", i.Name))
	}
}

File: MicroPanda/AST/Expression/Identifier.cs
namespace MicroPanda.AST.Expression;

using Type;

internal abstract class Identifier : Expression
{
    internal string? Name { get; set; }
    internal string? Qualified { get; set; }
    internal bool IsNamespace { get; set; }

    internal override void Validate(Context context, Type? expected)
    {
        var type = context.FindObject(Name!);
        if (t == null)
        {
            var (found, d) = c.Program.FindMember(Name);
            if (d is Variable v)
            {
                Const = v.Const;
                Typ = v.Type;
                Qualified = d.QualifiedName();
            }
            else if (d is Function f)
            {
                Const = true;
                Typ = f.Type;
                Qualified = d.QualifiedName();
            }
            else if (d is Enum)
            {
                Qualified = d.QualifiedName();
            }
            else if (d == null)
            {
                IsNamespace = c.Program.IsNamespace(Name);
            }
        }
        else
        {
            Const = false;
            Typ = t;
        }

        // * type would be null for enum (its member has type u8)
        // * type is null when identifier is namespace
        if (Typ == null && string.IsNullOrEmpty(Qualified) && !IsNamespace)
        {
            c.Program.Error(Position, $"undefined {Name}");
        }
    }
}